thetis/timeintegrator.py

```diff
@@ -566,6 +834,73 @@ def advance_linear(self, t, dt, solution, update_forcings):
             self.funcs_old[k].assign(self.funcs[k])
 
 
+class CrankNicolsonNew(TimeIntegrator):
+    """Standard Crank-Nicolson time integration scheme."""
+    def __init__(self, equation, solution, fields, dt, bnd_conditions=None, solver_parameters={}, gamma=0.5):
+        """Creates forms for the time integrator"""
+        super(CrankNicolsonNew, self).__init__(equation, solver_parameters)
+        self.solver_parameters.setdefault('snes_monitor', False)
+        self.solver_parameters.setdefault('snes_type', 'newtonls')
+
+        self.dt_const = Constant(dt)
+
+        self.solution = solution
+        self.solution_old = Function(self.equation.function_space)
+        self.fields = fields
+        # create functions to hold the values of previous time step
+        # TODO is this necessary? is self.fields sufficient?
+        self.fields_old = {}
+        for k in self.fields:
+            if self.fields[k] is not None:
+                if isinstance(self.fields[k], Function):
+                    self.fields_old[k] = Function(
+                        self.fields[k].function_space())
+                elif isinstance(self.fields[k], Constant):
+                    self.fields_old[k] = Constant(self.fields[k])
+
+        u = self.solution
+        u_old = self.solution_old
+        bnd = bnd_conditions
+        f = self.fields
+        f_old = self.fields_old
+
+        # Crank-Nicolson
+        gamma_const = Constant(gamma)
+        # FIXME this is consistent with previous implementation but time levels are incorrect
+        self.F = (self.equation.mass_term(u) - self.equation.mass_term(u_old) -
+                  self.dt_const*(gamma_const*self.equation.get_residual('all', u, u, f, f, bnd) +
+                                 (1-gamma_const)*self.equation.get_residual('all', u_old, u_old, f_old, f_old, bnd)
```

Comment:
So if you were using the same field values in `f` and `f_old` this could be simpler, you would just write:

```python
u = gamma_const*u + (1 - gamma_const)*u_old
self.equation.get_residual('all', u, u, f_old, f_old, bnd)
```
This gives you a residual that has C-N splitting for the solution and lags all the fields.  If one wants to use old and new field values as well as implicit solution values in the lagged `u`, we might want to think about redefining get_residual to take something other than a dict of fields.